### ISSUE TITLE
"database query error" with event trigger when mutation happens through v1/query or v2/query

### ISSUE THREAD

**Issue by deathemperor** [hasura/graphql-engine#7087]
```json
{
	"type": "http-log",
	"timestamp": "2021-06-19T18:33:42.160+0000",
	"level": "error",
	"detail": {
		"operation": {
			"user_vars": {
				"x-hasura-role": "admin"
			},
			"error": {
				"internal": {
					"statement": "\n           SELECT hdb_catalog.insert_event_log($1, $2, $3, $4, $5)\n                ",
					"prepared": true,
					"error": {
						"exec_status": "FatalError",
						"hint": null,
						"message": "invalid input syntax for type json",
						"status_code": "22P02",
						"description": "The input string ended unexpectedly."
					},
					"arguments": ["(Oid 25,Just (\"public\",Binary))", "(Oid 25,Just (\"medical_providers\",Binary))", "(Oid 25,Just (\"MP\",Binary))", "(Oid 25,Just (\"MANUAL\",Binary))", "(Oid 114,Just (\"{\\\"row\\\":{\\\"is_direct_billing\\\":false,\\\"email\\\":null,\\\"status\\\":0,\\\"phone\\\":null,\\\"expired_at\\\":null,\\\"latitude\\\":\\\"10.68894\\\",\\\"address\\\":null,\\\"is_excluded\\\":false,\\\"name\\\":\\\"an giang\\\",\\\"medical_provider_id\\\":\\\"130a52ef-e155-4879-9a67-5c98672917c8\\\",\\\"updated_at\\\":\\\"2021-06-19T18:02:17.232539+00:00\\\",\\\"deleted_at\\\":null,\\\"geocode_data\\\":{\\\"Municipality\\\":\\\"Ch\\195\\162u \\196\\144\\225\\187\\145c\\\",\\\"Country\\\":\\\"VNM\\\",\\\"SubRegion\\\":\\\"An Giang\\\",\\\"Geometry\\\":{\\\"Point\\\":[105.148,10.68894]},\\\"Label\\\":\\\"B\\225\\187\\135nh Vi\\225\\187\\135n \\196\\144a Khoa Khu V\\225\\187\\177c T\\225\\187\\137nh An Giang, Ch\\195\\162u \\196\\144\\225\\187\\145c, Vi\\225\\187\\135t Nam\\\",\\\"Neighborhood\\\":\\\"Ph\\198\\176\\225\\187\\157ng V\\196\\169nh M\\225\\187\\185\\\"},\\\"province\\\":null,\\\"created_at\\\":\\\"2021-06-19T17:54:08.038602+00:00\\\",\\\"type\\\":\\\"BVDK_C\\\",\\\"warning_level\\\":0,\\\"longitude\\\":\\\"105.148\\\",\\\"description\\\":null}}\",Binary))"]
				},
				"path": "$.args",
				"error": "database query error",
				"code": "unexpected"
			},
			"request_id": "08561e4b-f940-44c4-af1f-22159fcd051a",
			"response_size": 1643,
			"query": {
				"resource_version": null,
				"args": {
					"payload": {
						"row": {
							"is_direct_billing": false,
							"email": null,
							"status": 0,
							"phone": null,
							"expired_at": null,
							"latitude": "10.68894",
							"address": null,
							"is_excluded": false,
							"name": "an giang",
							"medical_provider_id": "130a52ef-e155-4879-9a67-5c98672917c8",
							"updated_at": "2021-06-19T18:02:17.232539+00:00",
							"deleted_at": null,
							"geocode_data": {
								"Municipality": "Châu Đốc",
								"Country": "VNM",
								"SubRegion": "An Giang",
								"Geometry": {
									"Point": [105.148, 10.68894]
								},
								"Label": "Bệnh Viện Đa Khoa Khu Vực Tỉnh An Giang, Châu Đốc, Việt Nam",
								"Neighborhood": "Phường Vĩnh Mỹ"
							},
							"province": null,
							"created_at": "2021-06-19T17:54:08.038602+00:00",
							"type": "BVDK_C",
							"warning_level": 0,
							"longitude": "105.148",
							"description": null
						}
					},
					"name": "MP",
					"source": "default"
				},
				"version": 1,
				"type": "pg_invoke_event_trigger"
			}
		},
		"request_id": "08561e4b-f940-44c4-af1f-22159fcd051a",
		"http_info": {
			"status": 500,
			"http_version": "HTTP/1.1",
			"url": "/v1/metadata",
			"ip": "172.19.0.1",
			"method": "POST",
			"content_encoding": null
		}
	}
}
```

**Comment by deathemperor** [comment]
I believe it belongs to this, or at least relates to: When inserting a new record to a table with JSONB fields:

```sql
{
    "statement": "INSERT INTO \"public\".\"medical_providers\" (\"name\", \"type\", \"geocode_data\")\nVALUES ('test', 'PKCK', '{}');",
    "prepared": false,
    "error": {
        "exec_status": "FatalError",
        "hint": null,
        "message": "invalid input syntax for type json",
        "status_code": "22P02",
        "description": "The input string ended unexpectedly."
    },
    "arguments": []
}
```

Got the same error when omitting `geocode_data` from query or setting it to null

**Comment by kedarguy** [comment]
I'm also getting this in v2.0.1

**Comment by albttx** [comment]
Same issue

**Comment by anttiviljami** [comment]
Same issue 👍

```
2021-07-08T14:57:45.221Z	b809d490-cdc5-4af7-aaa5-47a487ff344b	ERROR	{
    "message": "invalid input syntax for type json",
    "length": 499,
    "name": "error",
    "severity": "ERROR",
    "code": "22P02",
    "detail": "The input string ended unexpectedly.",
    "where": "JSON data, line 1: \nPL/pgSQL function hdb_catalog.insert_event_log(text,text,text,text,json) line 12 at assignment\nSQL statement \"SELECT hdb_catalog.insert_event_log(CAST('public' AS text), CAST('organization' AS text), CAST('audit_organization' AS text), TG_OP, _data)\"\nPL/pgSQL function hdb_catalog.\"notify_hasura_audit_organization_INSERT\"() line 25 at PERFORM",
    "file": "json.c",
    "line": "1137",
    "routine": "report_parse_error",
    "errorName": "error",
    "errorMessage": "invalid input syntax for type json",
    "stackTrace": "error: invalid input syntax for type json\n    at Parser.parseErrorMessage (/var/task/node_modules/pg-protocol/src/parser.ts:357:11)\n    at Parser.handlePacket (/var/task/node_modules/pg-protocol/src/parser.ts:186:21)\n    at Parser.parse (/var/task/node_modules/pg-protocol/src/parser.ts:101:30)\n    at TLSSocket.<anonymous> (/var/task/node_modules/pg-protocol/src/index.ts:7:48)\n    at TLSSocket.emit (events.js:314:20)\n    at TLSSocket.EventEmitter.emit (domain.js:483:12)\n    at addChunk (_stream_readable.js:297:12)\n    at readableAddChunk (_stream_readable.js:272:9)\n    at TLSSocket.Readable.push (_stream_readable.js:213:10)\n    at TLSWrap.onStreamRead (internal/stream_base_commons.js:188:23)",
    "awsRegion": "eu-central-1",
    "functionName": "pgdb-helper-staging-post",
    "functionVersion": "$LATEST",
    "functionMemorySize": "256",
    "level": 50,
    "sLevel": "ERROR"
}
```

**Comment by frandiox** [comment]
I can see this error after upgrading to Hasura 2.0.1 (from 1.3.3) when using raw SQL statements (without even retrying events):

```
"statement": "\n          INSERT INTO order_fees(order_id, type_id, amount_usd)\n          VALUES ('f75ce340-af2d-42bd-96a8-1dc91c4d4d05', 'PERCENT', 557), ('f75ce340-af2d-42bd-96a8-1dc91c4d4d05', 'MINIMUM', 0)\n          ON CONFLICT (order_id, type_id) WHERE item_id IS null DO UPDATE SET amount_usd = EXCLUDED.amount_usd;\n        ",
"prepared": false,
"error": {
  "exec_status": "FatalError",
  "hint": null,
  "message": "invalid input syntax for type json",
  "status_code": "22P02",
  "description": "The input string ended unexpectedly."
},
"arguments": []
```

The SQL statement is like this:

```sql
INSERT INTO order_fees(order_id, type_id, amount_usd)
VALUES ('f75ce340-af2d-42bd-96a8-1dc91c4d4d05', 'PERCENT', 557), ('f75ce340-af2d-42bd-96a8-1dc91c4d4d05', 'MINIMUM', 0)
ON CONFLICT (order_id, type_id) WHERE item_id IS null DO UPDATE SET amount_usd = EXCLUDED.amount_usd;
```

It works if I paste this code in the console directly, but not from the API. I've tried using the deprecated `/v1/query` and the new `/v2/query` with `run_sql` type.

**Comment by maguszoldik** [comment]
I confirm that we encounter the same error after migrating to 2.0.1, from the console there is a "database query error" after clicking on the invoke button. Same bad result when using the "invoke_event_trigger" api, which worked in 1.3.x  
It occurs even after deleting and fully recreating the event.  

"Not manual" events still continue to work as expected.

**Comment by frandiox** [comment]
@Varun-Choudhary @rikinsk  Hi! [This problem](https://github.com/hasura/graphql-engine/issues/7087#issuecomment-876646816) will prevent us from deploying our production app in a few weeks so I just wanted to ask: do you think this bug will be handled soonish or should I start adding `pg` client + DB credentials in my Hasura events for raw SQL access? Thanks!

**Comment by ifokeev** [comment]
same problem here

**Comment by rikinsk** [comment]
we are investigatting this folks

**Comment by tirumaraiselvan** [comment]
We know the root cause and will fix it soon.

**Comment by anttiviljami** [comment]
Any updates on this @tirumaraiselvan @kolharsam? Thanks in advance! <3

**Comment by rossng** [comment]
Also looks like we're experiencing this. Here's what comes out of the Postgres logs:

```
[DATABASE] [51-1]  sql_error_code = 22P02 ERROR:  invalid input syntax for type json
[DATABASE] [51-2]  sql_error_code = 22P02 DETAIL:  The input string ended unexpectedly.
[DATABASE] [51-3]  sql_error_code = 22P02 CONTEXT:  JSON data, line 1:
[DATABASE] [51-4] 	PL/pgSQL function hdb_catalog.insert_event_log(text,text,text,text,json) line 13 at assignment
[DATABASE] [51-5]  sql_error_code = 22P02 STATEMENT:
[DATABASE] [51-6] 	    SELECT hdb_catalog.insert_event_log($1, $2, $3, $4, $5)
[DATABASE] [51-7]
```

**Comment by Maxhodges** [comment]
Hey we can't launch in Production because of this issue. We "upgraded" to just before moving to production and now we can't launch for a month now. Can you guy make this a priority (or invest to give us some more runway?)

**Comment by Maxhodges** [comment]
> We know the root cause and will fix it soon.

26 days

**Comment by ifokeev** [comment]
yeah we can't upgrade too

**Comment by kolharsam** [comment]
Hello folks, thank you for reporting the issue, and this has been addressed in the latest release of the [graphql-engine (v2.0.7)](https://github.com/hasura/graphql-engine/releases/tag/v2.0.7). Apologies for the delay and any inconvenience caused.

**Comment by frandiox** [comment]
@kolharsam I've tried 2.0.7 and our issue still happens (exactly the same error). See details in [this comment](https://github.com/hasura/graphql-engine/issues/7087#issuecomment-876646816).

The image I'm using in a local Docker is `hasura/graphql-engine:v2.0.7.cli-migrations-v3`.

**Comment by kolharsam** [comment]
@frandiox, thanks for highlighting this again. We shall take a better look and get back to you soon.

cc @scriptonist @scriptnull

**Comment by tirumaraiselvan** [comment]
@frandiox Just to be clear, this is not an issue with Event Trigger and manually invoking an event via the console but an issue with you insert statement as described in your [comment](https://github.com/hasura/graphql-engine/issues/7087#issuecomment-876646816) :

```
INSERT INTO order_fees(order_id, type_id, amount_usd)
VALUES ('f75ce340-af2d-42bd-96a8-1dc91c4d4d05', 'PERCENT', 557), ('f75ce340-af2d-42bd-96a8-1dc91c4d4d05', 'MINIMUM', 0)
ON CONFLICT (order_id, type_id) WHERE item_id IS null DO UPDATE SET amount_usd = EXCLUDED.amount_usd;
```

This could be a different issue and I am trying to replicate it. Can you share the full schema of your `order_fees` table including any unique/exclusion constraints etc.

**Comment by tirumaraiselvan** [comment]
@frandiox Just to be clear, the INSERT statement throws that error only when the table has an event trigger defined, correct?

**Comment by frandiox** [comment]
@tirumaraiselvan Yes, I think you're correct. I tried removing the events from the table and then the raw SQL started working. Reverted the whole thing and it started failing again.

**Comment by tirumaraiselvan** [comment]
@frandiox Thanks for the confirmation. The original issue was for manual event trigger which has been fixed in 2.0.7

We have found another issue when mutation happens through v1/query or v2/query which is related to your comment. We will fix this ASAP. Sorry for the inconvenience.

**Comment by maguszoldik** [comment]
Hi.

I updated my version to 2.0.7 to do some tests.
Now it works from the console, but I still can't use it through the v1/query, with an error that differs from what you describe.

Maybe it is just the documentation that is outdated, I followed the API way described in this page
https://hasura.io/docs/latest/graphql/core/event-triggers/invoke-trigger-manually.html#invoke-trigger-manually

In the console, the event appears named as `eav_subcategory`.  
But I get the answer `{"path":"$.args","error":"event trigger \"eav_subcategory\" does not exist","code":"not-exists"}`

It may be related, the thing that has changed between my 1.3 and 2.0 usage is that I now use 2 databases connections, named `cms` and `erp`. The event I target is in `erp`.

**Comment by joserocha3** [comment]
> @frandiox Thanks for the confirmation. The original issue was for manual event trigger which has been fixed in 2.0.7
> 
> We have found another issue when mutation happens through v1/query or v2/query which is related to your comment. We will fix this ASAP. Sorry for the inconvenience.

Hi @tirumaraiselvan, any news on when the fix for the issue @frandiox described will get implemented and released?

Thanks!

**Comment by tirumaraiselvan** [comment]
@joserocha3 @frandiox Yes, we have a fix in progress and expect to land in few days. I will re-open this issue and change the title to reflect the same.

**Comment by Maxhodges** [comment]
> We know the root cause and will fix it soon.

47 days

**Comment by tirumaraiselvan** [comment]
This has now been fixed in 2.0.9 and rollout of the same version is happening on Cloud as well.

Sorry for the delay on this, there were 2 issues in this report and the second one turned out to be quite complicated.

**Comment by Shalashtein** [comment]
Getting similar errors after my upgrade from 1.3.3 to 2.0.10, I had exposed the the hbd_catalog.event_log in a VIEW and while starting up hasura runs some commands to ALTER the LOCKED column on hbd_catalog.event_log which fails because the VIEW depends on it. Deleted the view and now I'm getting this in my ECS logs:

```
{"type":"event-trigger","timestamp":"2021-11-22T07:11:53.187+0000","level":"error","detail":{"internal":{"statement":"\n UPDATE hdb_catalog.event_log\n SET locked = 't'\n WHERE id IN ( SELECT l.id\n FROM hdb_catalog.event_log l\n WHERE l.delivered = 'f' and l.error = 'f' and l.locked = 'f'\n and (l.next_retry_at is NULL or l.next_retry_at <= now())\n and l.archived = 'f'\n ORDER BY created_at\n LIMIT $1\n FOR UPDATE SKIP LOCKED )\n RETURNING id, schema_name, table_name, trigger_name, payload::json, tries, created_at\n ","prepared":true,"error":{"exec_status":"FatalError","hint":null,"message":"invalid input syntax for type timestamp with time zone: \"f\"","status_code":"22007","description":null},"arguments":["(Oid 20,Just (\"\\NUL\\NUL\\NUL\\NUL\\NUL\\NUL\\NULd\",Binary))"]},"path":"$","error":"database query error","code":"unexpected"}}
```

**Comment by Edmartt** [comment]
I'm still having this issue in version 2.0.10 in any moment. If I do a mutations or modify something inside the console is the same, I need to reload like I'm doing from outside, this is not supposed to happen

**Comment by sasog23** [comment]
v2.5.0

still getting an error "invalid input syntax for type json" on update
if I remove event trigger, than there is no error

**Comment by sasog23** [comment]
[v2.12.0] i'm still getting this error...any news ?
i have an event trigger on the table servicedesk.communication

pgSQL function hdb_catalog.insert_event_log(text,text,text,text,json) line 12 at assignment
SQL statement \"SELECT hdb_catalog.insert_event_log(CAST('servicedesk' AS text), CAST('communication' AS text)
, CAST('t_communication_send_email' AS text), TG_OP, _data)\"
PL/pgSQL function hdb_catalog.\"notify_hasura_t_communication_send_email_INSERT\"() line 25 at PERFORM
SQL statement \"INSERT INTO servicedesk.communication (helpdesk_id, detail_description, user_id, contact_id, process_definition_matrix_id, form_id)
        SELECT
		_hd_id,
		JSONB_BUILD_OBJECT('plain', p_body_plain, 'eml_id', p_eml_id),
		_user_id,
		_contact_id,
		_process_definition_matrix_id,
		CASE WHEN is_contact_agent THEN 170 ELSE 171 END
		RETURNING \"id\"\"
		PL/pgSQL function _mailer.account_37(jsonb) line 277 at SQL statement
		SQL statement \"CALL _mailer.account_37($1);\"
		PL/pgSQL function _mailer.main_proc(jsonb) line 21 at EXECUTE"

**Comment by CuriousLearner** [comment]
Hi @tirumaraiselvan @codingkarthik @kolharsam @rikinsk !

This is still happening on version 2.16.0 and we're unable to release on production. This does not happen for us on any other environments.

This is the log:

```
invalid input syntax for type json
DETAIL:  The input string ended unexpectedly.
CONTEXT:  JSON data, line 1: 
PL/pgSQL function hdb_catalog.insert_event_log(text,text,text,text,json) line 12 at assignment
SQL statement \"SELECT hdb_catalog.insert_event_log(CAST('public' AS text), CAST('action' AS text), CAST('actionChanges' AS text), TG_OP, _data)\"
PL/pgSQL function hdb_catalog.\"notify_hasura_actionChanges_DELETE\"() line 25 at PERFORM
```

Please note that I tried to delete these triggers and manually re-create them. It kind of resolved the issue -- but it started happening again out of the blue. 

The error happens sporadically for us, so we are unable to find the root cause. Please help. I've been debugging this for over a week and it's driving me crazy especially with a deadline to release in production.

**Comment by CuriousLearner** [comment]
Hi @meetzaveri , Can you please take a look at the above, once?

**Comment by Qasim-Aziz** [comment]
i have user table and using events to run when my laravel application trying to save user information it through error 

`Undefined function: 7 ERROR:  function hdb_catalog.insert_event_log(text, text, text, text, json) does not exist\nLINE 1: SELECT hdb_catalog.insert_event_log(CAST('public' AS text), ...\n               ^\nHINT:  No function matches the given name and argument types. You might need to add explicit type casts.\nQUERY:  SELECT hdb_catalog.insert_event_log(CAST('public' AS text), CAST('users' AS text), CAST('SendEmailVerificationNotification' AS text), TG_OP, _data)\nCONTEXT:  PL/pgSQL function hdb_catalog."notify_hasura_SendEmailVerificationNotification_INSERT"() line 29 at PERFORM (SQL: insert into "users"`

**Comment by jakubste** [comment]
I have similar problem. In our case it's a trigger triggered by remote schema mutation.

```
invalid input syntax for type json\nDETAIL:  The input string ended unexpectedly.\nCONTEXT:  JSON data, line 1: \nfunkcja PL/pgSQL hdb_catalog.insert_event_log(text,text,text,text,json), wiersz 12 w przypisanie\nSQL statement \"SELECT hdb_catalog.insert_event_log(CAST('public' AS text), CAST('rungs_userrung' AS text), CAST('updateUserRung' AS text), TG_OP, _data)\"\nfunkcja PL/pgSQL hdb_catalog.\"notify_hasura_updateUserRung_UPDATE\"(), wiersz 25 w PERFORM
```
